Dylanmccall/pygobject gtk

dylanmccall opened the pull request:
This branch adds a GTK front-end, with a webview provided by WebKitGTK. These libraries are available out of the box in the org.freedesktop.Platform Flatpak runtime, so an application using pyeverywhere can be packaged nicely on Linux with no additional dependencies.

dylanmccall committed:
Add a GTK front-end using pygobject

dylanmccall commented:
Hi @kollivier ! I'm happy with this now if you'd like to take another look. Added support for a menubar so the existing API is (hopefully) all supported :) 

Here's a screenshot with the existing Kolibri Mac app:

![Screenshot from 2020-03-23 17-48-19](https://user-images.githubusercontent.com/132063/77376843-98574200-6d2e-11ea-877a-7715ffa08d3d.png)

A Linux app for Kolibri should be able to use mostly the same code as the Mac app. Probably would skip the Edit menu there (those items aren't hooked up to anything with my example, and the keyboard shortcuts for Undo / Redo / Copy & Paste work automatically). In addition, the menu shortcuts (passed as a specially formatted string) would need to be specified in a [slightly different format](https://developer.gnome.org/gtk3/stable/gtk3-Keyboard-Accelerators.html#gtk-accelerator-parse) than with wxWidgets.

kollivier commented:
Awesome, thanks! A bit busy finishing up some work for a project, but wanted to quickly address a couple of the points you made. 

We would be fine with having the Kolibri app UX customized in appropriate ways for each platform. You can add `if pew.ui.platform == 'gtk':` for any UI items that should, or should not, only appear on the GTK platform.

For the menu items, I'm wondering if you think it would make sense to either auto-translate the menu shortcuts (e.g. convert the `CTRL+` to `<CTRL>` in the GTK platform code) or devise our own syntax at the PyEverywhere level (e.g. `PEWShortcut(keys=['C'], modifiers=['CTRL'])` and then have each platform convert to native format. I'd be open to either way, though my preference might be for the latter. Main thing I want to avoid is to have to specify the shortcuts differently in client code for each platform. What do you think? Next week I should hopefully have some time to help out with this part.

dylanmccall commented:
Yeah, the second approach to shortcuts (devising our own syntax) would make plenty of sense to me. I was thinking about adding something like that here, even, but didn't want to put too much stuff in this branch :) It would fit into the way GTK handles those: accelerators / shortcuts in GTK are specified as an accelerator key (with a key code), and an array of modifier keys (from a limited set). So all we'd have to do is translate key codes, and possibly throw an exception or print a warning if a modifier from PEWShortcut isn't applicable to GTK.

pwithnall commented on src/pew/pygobject_gtk/__init__.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def run_on_main_thread(func, *args, **kwargs):
+    return func(*args, **kwargs)
```
I don’t know any of the context for this, but from a Python/GLib point of view, I don’t think this will run something on the main thread — I think it’ll just run `func` on the current thread.

To call a function on the main thread using the GLib main loop, use `GLib.MainContext.invoke(None, func, data)` (you might need to marshal `*args` and `**kwargs` into `data` and then call `func` from a wrapper function which demarshals them again). The wrapper function should return `False` (see [the documentation](https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html#g-main-context-invoke)).

dylanmccall replied:
Thank you! Definitely seems it should be doing that :)

dylanmccall committed:
Use GLib.MainContext.invoke_full for run_on_main_thread

kollivier commented on setup.py:
```diff
@@ -17,8 +17,9 @@
-        ':platform_machine=="x86_64"': 'wxPython',
-        ':sys_platform=="darwin"': 'dmgbuild',
+        ':sys_platform=="darwin"': ['dmgbuild'],
+        ':sys_platform=="darwin" and platform_machine=="x86_64"': ['wxPython'],
```
Windows is the only platform that currently requires wxPython, as Mac (and now GTK!) have alternate options, so maybe we should switch this to Windows? One piece I need to think through more is how best to streamline the backend setup. On platforms like Linux, you could setup Android, GTK, or wx backends (or multiple), so probably best to let the user pick and not try to install all of those from setup.py.

For now I guess we can do it through docs, but it'd be nice if there was somehow a way to script or automate the setup of a backend.

kollivier commented on src/pew/pygobject_gtk/__init__.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def choose_file(callback):
+    print("choose_file", callback)
```
Could we use `logging.warning` statements instead, so that when running as a GUI these would get logged?

dylanmccall replied:
Oh, sure! And I just realized I forgot to implement that, didn't I?

kollivier replied:
It's okay with me to leave some things unimplemented, especially if you have no use case yourself or no easy way to test. As long as the logs warn about it, I think that's good. I'd also be okay with raising `NotImplementedError`.

kollivier reviewed (changes requested):
Looks great, and the Mac app code ran on GTK with this PR and everything I tested worked well! The one thing I think that we should address I think we should either adjust or remove the `extra_requires` dependencies that are backend specific and then we'll be ready to go.

dylanmccall committed:
Implement file dialogs for the gtk frontend

dylanmccall committed:
Implement application registration with Gtk.Application

This adds an optional application_id field in PEWApp. If this is set to
some well-known string in a subclass, GTK will ensure that only one
instance with that application_id is running. Applications using
pyeverywhere with GTK can set application_id to a value such as
"com.example.Application" to support this feature.

dylanmccall committed:
List optional dependencies without installing them

dylanmccall commented:
Okay, I changed the format of extras_require a little so it says what dependencies are required for certain optional features without causing them to be installed. Removed some `print` statements (by implementing the unimplemented functions they were denoting and I forgot about :b). I also cleaned up the Gtk.Application stuff a little. I can do that shortcuts stuff in another branch when we do the Kolibri Linux installer repo :)

kollivier reviewed (approved):
Looking good, thanks a lot for this!
